skimage/util/_flood_fill_cy.pyx

```diff
@@ -0,0 +1,190 @@
+
+
+ctypedef fused dtype_t:
+    cnp.uint8_t
+    cnp.uint16_t
+    cnp.uint32_t
+    cnp.uint64_t
+    cnp.int8_t
+    cnp.int16_t
+    cnp.int32_t
+    cnp.int64_t
+    cnp.float32_t
+    cnp.float64_t
+
+
+# Definition of flag values used for `flags` in _local_maxima & _fill_plateau
+cdef:
+    # Border value - do not cross!
+    unsigned char BORDER = 2
+    # Part of the flood fill
+    unsigned char FILL = 1
+    # Not checked yet
+    unsigned char UNKNOWN = 0
+
+
+def _flood_fill(dtype_t[::1] image not None,
+                unsigned char[::1] flags,
+                Py_ssize_t[::1] neighbor_offsets not None,
+                Py_ssize_t ravelled_seed_idx,
+                dtype_t seed_value,
+                unsigned char do_tol,
+                dtype_t high_tol,
+                dtype_t low_tol):
+    """Find the region to be filled.
+
+    Inner function to `flood_fill` that detects all connected points equal (or
+    within tolerance) of the initial seed point.  The result is stored inplace
+    inside `flags` as the value "FILL".
+
+    Parameters
+    ----------
+    image : ndarray, one-dimensional
+        The raveled view of a n-dimensional array.
+    flags : ndarray
+        An array of flags that is used to store the state of each pixel during
+        evaluation and is MODIFIED INPLACE. Initially, pixels that border the
+        image edge must be marked as "BORDER_INDEX" while all other pixels
+        should be marked with "EXCLUDED".
+    neighbor_offsets : ndarray
+        A one-dimensional array that contains the offsets to find the
+        connected neighbors for any index in `image`.
+    ravelled_seed_idx : int
+        Ravelled seed index for flood filling.
+    do_tol : unsigned char
+        If 0, only exactly equal connected points are filled and `tolerance` is
+        ignored (fastest).  Otherwise, points within `tolerance` of start value
+        will also be filled (inclusive).
+    high_tol :
+        Upper limit for tolerance comparison (ignored if `do_tol` is 0).
+    low_tol :
+        Lower limit for tolerance comparison (ignored if `do_tol` is 0).
+    """
+    cdef:
+        QueueWithHistory queue
+        unsigned char last_dim_in_neighbors
+
+    with nogil:
+        # Initialize a buffer used to queue positions while evaluating each
+        # potential maximum (flagged with 2)
+        queue_init(&queue, 64)
+        try:
+            # Conduct flood fill from this point - with or without tolerance
+            if do_tol == 0:
+                _flood_fill_do_equal(image, flags, neighbor_offsets, &queue,
+                    ravelled_seed_idx, seed_value)
+            else:
+                _flood_fill_do_tol(image, flags, neighbor_offsets, &queue,
+                    ravelled_seed_idx, seed_value, high_tol, low_tol)
+        finally:
+            # Ensure that memory is released again
+            queue_exit(&queue)
+
+
+cdef inline void _flood_fill_do_equal(
```

Comment:
Just a minor suggestion: How about naming this one `_fill_plateau` or `_fill_equal` and the other one below `_fill_with_tolerance`? But this is not a strong opinion.